ui/src/app/services/workspace-storage.service.ts

```diff
@@ -17,83 +13,43 @@ export interface WorkspaceData extends Workspace {
 
 @Injectable()
 export class WorkspaceStorageService {
-  private activeCall = false;
-  private activeCallObservable: Observable<WorkspaceResponse>;
-  private loadingWorkspace = new ReplaySubject<WorkspaceData>(1);
-  private cachedWorkspace: WorkspaceData = {
-    accessLevel: WorkspaceAccessLevel.NOACCESS,
-    name: '',
-    researchPurpose: {
-      diseaseFocusedResearch: false,
-      methodsDevelopment: false,
-      controlSet: false,
-      aggregateAnalysis: false,
-      ancestry: false,
-      commercialPurpose: false,
-      population: false,
-      reviewRequested: false,
-    }
-  };
-
-  public loadingWorkspace$ = this.loadingWorkspace.asObservable();
+  // Cache of loading or completed Promises for workspace data. Key is of the
+  // form "ns/id".
+  private cache = new Map<string, Promise<WorkspaceData>>();
```

Comment:
So, is there a concern around shared workspaces for this? I think that’s my one concern, that a map increases the chance that you open one workspace, switch to another and work on that one for a while, and meanwhile someone works on workspace a.

Reply:
Good point - I don't think is specific to multiple workspaces though. The same staleness issue could happen within the context of one workspace.

Not sure if we need to worry about this now, but if it's a concern we could turn this into an timed expiring cache.

Reply:
Yeah, maybe we just do it as is for now and then follow up if it becomes a problem.